### PR TITLE
[SPARK-39218][SS][PYTHON] Make foreachBatch streaming query stop gracefully

### PR DESCRIPTION
**Pull request by HyukjinKwon**
### What changes were proposed in this pull request?

This PR proposes to make the `foreachBatch` streaming query stop gracefully by handling the interrupted exceptions at `StreamExecution.isInterruptionException`.

Because there is no straightforward way to access to the original JVM exception, here we rely on string pattern match for now (see also "Why are the changes needed?" below). There is only one place from Py4J https://github.com/py4j/py4j/blob/master/py4j-python/src/py4j/protocol.py#L326-L328 so the approach would work at least.

### Why are the changes needed?

In `foreachBatch`,  the Python user-defined function in the microbatch runs till the end even when `StreamingQuery.stop` is invoked. However, when any Py4J access is attempted within the user-defined function:

- With the pinned thread mode disabled, the interrupt exception is not blocked, and the Python function is executed till the end in a different thread.
- With the pinned thread mode enabled, the interrupt exception is raised in the same thread, and the Python thread raises a Py4J exception in the same thread.

The latter case is a problem because the interrupt exception is first thrown from JVM side (`java.lang. InterruptedException`) -> Python callback server (`py4j.protocol.Py4JJavaError`) -> JVM (`py4j.Py4JException`), and `py4j.Py4JException` is not listed in `StreamExecution.isInterruptionException` which doesn't gracefully stop the query.

Therefore, we should handle this exception at `StreamExecution.isInterruptionException`.

### Does this PR introduce _any_ user-facing change?

Yes, it will make the query gracefully stop.

### How was this patch tested?

Manually tested with:

```python
import time

def func(batch_df, batch_id):
    time.sleep(10)
    print(batch_df.count())

q = spark.readStream.format("rate").load().writeStream.foreachBatch(func).start()
time.sleep(5)
q.stop()
```

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala [hunk 1]
```diff
@@ -647,6 +655,10 @@ object StreamExecution {
       } else {
         false
       }
+    // py4j.Py4JException - with pinned thread mode on, the exception can be interrupted by Py4J
+    //                      access, for example, in `DataFrameWriter.foreachBatch`. See also
+    //                      SPARK-39218.
+    case e: py4j.Py4JException => PROXY_ERROR.findFirstIn(e.getMessage).isDefined
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala:661]: To fully address, we should have a way to pass the original exception around .. but seems not easy at this moment. This is more just a workaround for now.

**Comment by HyukjinKwon** [comment]
cc @ueshin @zsxwing FYI

**Comment by zsxwing** [comment]
@HyukjinKwon Thanks for fixing this. Could you write a python test like https://github.com/apache/spark/blob/master/sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamSuite.scala#L1172-L1178 ? This test was written before we supported foreach. Now with `foreachBatch`, we should be able to throw an exception easily using
```
def runBatch(df, id):
   use py4j to throw a jvm exception
df.writeStream.foreachBatch(runBatch).start()
```

File: python/pyspark/sql/tests/test_streaming.py [hunk 2]
```diff
@@ -592,6 +592,18 @@ def collectBatch(df, id):
...
+        # SPARK-39218: Make foreachBatch streaming query stop gracefully
+        def func(batch_df, _):
+            time.sleep(10)
+            batch_df.count()
+
+        q = self.spark.readStream.format("rate").load().writeStream.foreachBatch(func).start()
+        time.sleep(5)
+        q.stop()
+        time.sleep(15)  # Wait enough for the exception to be propagated if exists.
+        self.assertIsNone(q.exception(), "No exception has to be propagated.")
```
Comment (HyukjinKwon) [inline python/pyspark/sql/tests/test_streaming.py:605]: I just took a simpler path. I can use the test source approach like `ThrowingExceptionInCreateSource` but that would require more complicated test cases like https://github.com/apache/spark/blob/master/python/pyspark/sql/tests/test_dataframe.py#L1245-L1266.

The test cases here allow false positive but won't be flaky at least.

File: python/pyspark/sql/tests/test_streaming.py [hunk 3]
```diff
@@ -592,6 +592,18 @@ def collectBatch(df, id):
             if q:
                 q.stop()
 
+    def test_streaming_foreachBatch_graceful_stop(self):
+        # SPARK-39218: Make foreachBatch streaming query stop gracefully
+        def func(batch_df, _):
+            time.sleep(10)
```
Comment (zsxwing) [inline python/pyspark/sql/tests/test_streaming.py:598]: How does this func trigger InterruptedException? I would expect codes like `self.spark._jvm.java.lang.Thread.sleep(10000)` instead.

Comment (HyukjinKwon) [reply python/pyspark/sql/tests/test_streaming.py:598]: Actually, `batch_df.count()` below triggers it via accessing to the interrupted Java thread if I'm not mistaken.

Comment (zsxwing) [reply python/pyspark/sql/tests/test_streaming.py:598]: I see. `batch_df.count()` probably will touch some code that can be interrupted. Can we use `self.spark._jvm.java.lang.Thread.sleep(10000)` to save 5 seconds in this test?

Comment (HyukjinKwon) [reply python/pyspark/sql/tests/test_streaming.py:598]: Yeah, that should work too.

Comment (HyukjinKwon) [reply python/pyspark/sql/tests/test_streaming.py:598]: Let me fix 👍

**Review by zsxwing — APPROVED** [review]
LGTM pending tests

**Comment by HyukjinKwon** [comment]
Merged to master and branch-3.3.

**Pull request closed without merging**